src/lib.rs

```diff
@@ -112,3 +115,60 @@ impl PrivilegeLevel {
         }
     }
 }
+
+/// A wrapper that can be used to safely create one mutable reference `&'static mut T` from a static variable.
+///
+/// `SingleUseCell` is safe because it ensures that it only ever gives out one reference.
+///
+/// ``SingleUseCell<T>` is a safe alternative to `static mut` or a static `UnsafeCell<T>`.
+#[derive(Debug)]
+pub struct SingleUseCell<T> {
+    used: AtomicBool,
+    value: UnsafeCell<T>,
+}
+
+impl<T> SingleUseCell<T> {
+    /// Construct a new SingleUseCell.
+    pub const fn new(value: T) -> Self {
+        Self {
+            used: AtomicBool::new(false),
+            value: UnsafeCell::new(value),
+        }
+    }
+
+    /// Try to acquire a mutable reference to the wrapped value.
+    /// This will only succeed the first time the function is
+    /// called and fail on all following calls.
+    ///
+    /// ```
+    /// use x86_64::SingleUseCell;
+    ///
+    /// static FOO: SingleUseCell<i32> = SingleUseCell::new(0);
+    ///
+    /// // Call `try_get_mut` for the first time and get a reference.
+    /// let first: &'static mut i32 = FOO.try_get_mut().unwrap();
+    /// assert_eq!(first, &0);
+    ///
+    /// // Calling `try_get_mut` again will return `None`.
+    /// assert_eq!(FOO.try_get_mut(), None);
+    /// ```
+    pub fn try_get_mut(&self) -> Option<&mut T> {
+        let already_used = self.used.swap(true, Ordering::AcqRel);
+        if already_used {
+            None
+        } else {
+            Some(unsafe {
+                // SAFETY: no reference has been given out yet and we won't give out another.
+                &mut *self.value.get()
+            })
+        }
+    }
+}
+
+// SAFETY: Sharing a `SingleUseCell<T>` between threads is safe regardless of whether `T` is `Sync`
+// because we only expose the inner value once to one thread. The `T: Send` bound makes sure that
+// sending a unique reference to another thread is safe.
+unsafe impl<T: Send> Sync for SingleUseCell<T> {}
+
```

Comment:
> because we only expose the inner value once to one thread.

I'm a newbee to rust. It looks a bit like ThreadLocal. Can you help me explain the difference

Reply:
I'm assuming you're referring to [`thread_local::ThreadLocal`](https://docs.rs/thread_local/1.1.3/thread_local/struct.ThreadLocal.html). `ThreadLocal` and `SingleUseCell` are completely different data structures: `ThreadLocal` returns a reference to a different object in each thread, while `SingleUseCell` only ever gives out a single reference regardless of what thread is calling the function.

If you had two thread's A and B, and you called `ThreadLocal::get` you might get a reference A on thread A and a reference B on thread B.

If you call `SingleUseCell::try_get_mut` for the first time, you'll get some reference, but after that if you'd call `SingleUseCell::try_get_mut` again it would return `None` regardless of which thread is executing.

Reply:
thank you